lib/mongo/collection/view/change_stream.rb

```diff
@@ -0,0 +1,138 @@
+# Copyright (C) 2017 MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'mongo/collection/view/change_stream/retryable'
+
+module Mongo
+  class Collection
+    class View
+
+      # Provides behaviour around a `$changeStream` pipeline stage in the
+      # aggregation framework. Specifying this stage allows users to request that
+      # notifications are sent for all changes to a particular collection or database.
+      #
+      # @note Only available in server versions 3.6 and higher.
+      #
+      # @since 2.5.0
+      class ChangeStream < Aggregation
+        include Retryable
+
+        # @return [ String ] The fullDocument option default value.
+        #
+        # @since 2.5.0
+        FULL_DOCUMENT_DEFAULT = 'default'.freeze
+
+        # @return [ BSON::Document ] The change stream options.
+        #
+        # @since 2.5.0
+        attr_reader :options
+
+        # Initialize the change stream for the provided collection view, pipeline
+        # and options.
+        #
+        # @example Create the new change stream view.
+        #   ChangeStream.new(view, pipeline, options)
+        #
+        # @param [ Collection::View ] view The collection view.
+        # @param [ Array<Hash> ] pipeline The pipeline of operators to filter the change notifications.
+        # @param [ Hash ] opts The change stream options.
+        #
+        # @option options [ String ] :full_document Allowed values: ‘default’, ‘updateLookup’. Defaults to ‘default’.
+        #   When set to ‘updateLookup’, the change notification for partial updates will include both a delta
+        #   describing the changes to the document, as well as a copy of the entire document that was changed
+        #   from some time after the change occurred.
+        # @option options [ BSON::Document, Hash ] :resume_after Specifies the logical starting point for the
+        #   new change stream.
+        # @option options [ Integer ] :max_await_time_ms The maximum amount of time for the server to wait
+        #   on new documents to satisfy a change stream query.
+        # @option options [ Integer ] :batch_size The number of documents to return per batch.
+        # @option options [ BSON::Document, Hash ] :collation The collation to use.
+        #
+        # @since 2.5.0
+        def initialize(view, pipeline, options = {})
+          @view = view
+          @change_stream_filters = pipeline
```

Comment:
Do you need to make a deep copy of the pipeline to ensure the user does not modify it after calling watch?
```ruby
pipeline = []
change_stream = coll.watch(pipeline, {})
pipeline << {'$project': {_id: 0}}
# Will the next cursor be created with [{'$changeStream': ...}, {'$project': {_id: 0}}]?
```
Maybe I'm being too paranoid...

Reply:
haha, that is possible, I guess. Good call though, it wouldn't hurt to copy it.